### PR TITLE
Updates documentation for tool suite experiments

### PR DESCRIPTION
**Pull request by vulder**
resolves se-passau/VaRA#572

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/experiments/wllvm.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5238778&bid=17171267)

File: docs/source/vara-ts-api/experiments.rst [hunk 1]
```diff
@@ -1,22 +1,42 @@
...
+``Experiments`` are the base concept how the tool suite and BenchBuild execute research experiments, e.g., measuring performance or analyzing a software project.
+They are designed to make the execution of research experiments easy and reproducable by separating experiment specific steps from project specific ones.
+For example, how a specific project is compiled is the responsibilitty of the project writer, where how the project is evaluated during a research experiment is the ``Experiment``'s task.
 
+.. note::
+  Details on how to run experiments can be found :ref:`here<Running experiments with BenchBuild>`
+
+Designing a new ``Experiment`` is also quite simple.
+
+* First, create a new python module in the experiments directory and add a experiment class which inherits from ``benchbuild.experiment.Experiment``.
```
Comment (boehmseb) [inline docs/source/vara-ts-api/experiments.rst]:
```suggestion
* First, create a new python module in the ``experiments`` directory and add an experiment class which inherits from ``benchbuild.experiment.Experiment``.
```

File: docs/source/vara-ts-api/experiments.rst [hunk 2]
```diff
@@ -1,22 +1,42 @@
...
+  If VaRA-TS should provide automatic support for analyzing different version, i.e., different revisions of a git based project, use :class:`~varats.utils.experiment_util.VersionExperiment` as base class.
```
Comment (boehmseb) [inline docs/source/vara-ts-api/experiments.rst]:
```suggestion
  If VaRA-TS should provide automatic support for analyzing different versions, i.e., different revisions of a git based project, use :class:`~varats.utils.experiment_util.VersionExperiment` as base class.
```

File: docs/source/vara-ts-api/experiments.rst [hunk 3]
```diff
@@ -1,22 +1,42 @@
...
+* Second, define two static variables for your experiment: ``NAME`` and ``REPORT_TYPE``
+* Next, override the ``actions_for_project`` method.
+  This method should assign run-time/compile-time extensions and specify the list of actions that should be performed.
+  Each action the experiment does is call a ``Step`` and will be executed by BenchBuild in order.
```
Comment (boehmseb) [inline docs/source/vara-ts-api/experiments.rst]:
```suggestion
  Each action the experiment does is called a ``Step`` and will be executed by BenchBuild in order.
```

File: docs/source/vara-ts-api/experiments.rst [hunk 4]
```diff
@@ -1,22 +1,42 @@
 Experiments
 ===========
 
-`TODO (se-passau/VaRA#572) <https://github.com/se-passau/VaRA/issues/572>`_: write docs
+``Experiments`` are the base concept how the tool suite and BenchBuild execute research experiments, e.g., measuring performance or analyzing a software project.
+They are designed to make the execution of research experiments easy and reproducable by separating experiment specific steps from project specific ones.
+For example, how a specific project is compiled is the responsibilitty of the project writer, where how the project is evaluated during a research experiment is the ``Experiment``'s task.
 
+.. note::
+  Details on how to run experiments can be found :ref:`here<Running experiments with BenchBuild>`
```
Comment (boehmseb) [inline docs/source/vara-ts-api/experiments.rst]:
```suggestion
  Details on how to run experiments can be found :ref:`here<Running experiments with BenchBuild>`.
```

File: docs/source/vara-ts-api/experiments.rst [hunk 5]
```diff
@@ -1,22 +1,42 @@
...
+* Second, define two static variables for your experiment: ``NAME`` and ``REPORT_TYPE``
+* Next, override the ``actions_for_project`` method.
+  This method should assign run-time/compile-time extensions and specify the list of actions that should be performed.
+  Each action the experiment does is call a ``Step`` and will be executed by BenchBuild in order.
+* Last, add your experiment for testing to the BenchBuild config file `vara-root/benchbuild/.benchbuild.yml` under plugins/experiments/value. After testing, integrate them into the tool suite by adding it to the experiment list in ``varats.settings.generate_benchbuild_config``, so it will be automatically added to the BenchBuild config in the future.
+
+.. note::
+  For more information about ``Experiment``'s consider reading the BenchBuild `docs <https://pprof-study.readthedocs.io/en/master/>`_.
+
+If you're looking for a simple example experiment, consider taking a look at :class:`~varats.experiments.just_compile.JustCompileReport`
```
Comment (boehmseb) [inline docs/source/vara-ts-api/experiments.rst]:
```suggestion
If you're looking for a simple example experiment, consider taking a look at :class:`~varats.experiments.just_compile.JustCompileReport`.
```

File: varats/experiments/blame_report_experiment.py [hunk 6]
```diff
@@ -1,8 +1,6 @@
 """
-Implements the blame report experiment.
-
-The experiment analyses a project with VaRAs blame analysis and generates a
-BlameReports.
+Implements the basic blame report experiment. The experiment analyses a project
+with VaRAs blame analysis and generates a BlameReports.
```
Comment (boehmseb) [inline varats/experiments/blame_report_experiment.py]:
```suggestion
with VaRA's blame analysis and generates a BlameReport.
```

File: docs/source/vara-ts/benchbuild.rst [hunk 7]
```diff
@@ -0,0 +1,24 @@
+Running experiments with BenchBuild
+===================================
+
+VaRA-TS provides different preconfigured experiments and projects.
+In order to execute an experiment on a project we use BenchBuild, an empirical-research toolkit.
```
Comment (boehmseb) [inline docs/source/vara-ts/benchbuild.rst]: Maybe link to BB github?

File: docs/source/vara-ts/benchbuild.rst [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+Setup: Configuring BenchBuild
+-----------------------------
+First, we need to generate a folder with a configuration file for BenchBuild in the vara root directory, this is done with:
```
Comment (boehmseb) [inline docs/source/vara-ts/benchbuild.rst]:
```suggestion
First, we need to generate a folder with a configuration file for BenchBuild in the vara root directory. This is done with:
```

File: docs/source/vara-ts/benchbuild.rst [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+.. code-block:: bash
+
+  vara-gen-bbconfig
+
+Running BenchBuild experiments
+----------------------------------
+Second, we change into the benchbuild folder and run an experiment that generates `BlameReports` for provided projects, in this case we use `gzip`.
```
Comment (boehmseb) [inline docs/source/vara-ts/benchbuild.rst]:
```suggestion
Second, we change into the benchbuild folder and run an experiment that generates `BlameReports` for provided projects. In this case we use `gzip`.
```

File: docs/source/vara-ts/benchbuild.rst [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+.. code-block:: bash
+
+  cd $VARA_ROOT/benchbuild
+  benchbuild -vv run -E GenerateBlameReport gzip
+
+The generated result files are place in the `vara/results/$PROJECT_NAME` folder and can be further visualized with VaRA-TS graph generators.
```
Comment (boehmseb) [inline docs/source/vara-ts/benchbuild.rst]:
```suggestion
The generated result files are place in the ``vara/results/$PROJECT_NAME`` folder and can be further visualized with VaRA-TS graph generators.
```

File: varats/experiments/blame_report_experiment.py [hunk 11]
```diff
@@ -114,7 +112,11 @@ class BlameReportExperiment(VersionExperiment):
 
     def actions_for_project(self, project: Project) -> tp.List[actions.Step]:
         """Returns the specified steps to run the project(s) specified in
```
Comment (boehmseb) [inline varats/experiments/blame_report_experiment.py]:
```suggestion
        """
        Returns the specified steps to run the project(s) specified in
```

File: varats/utils/experiment_util.py [hunk 12]
```diff
@@ -102,12 +104,19 @@ def __call__(self, ex: Exception) -> None:
...
+def get_default_compile_error_wrapped(
+        project: Project, report_type: tp.Type[BaseReport],
+        result_folder_template: str) -> FunctionPEErrorWrapper:
     """
     Setup the default project compile function with an error handler.
+
+    Args:
+        project: that will be compiled
+        report_type: that should be generated
+        result_folder_template: where the results would be placed
```
Comment (boehmseb) [inline varats/utils/experiment_util.py]:
```suggestion
        result_folder_template: where the results will be placed
```

**Review by boehmseb — CHANGES_REQUESTED** [review]

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/experiments/wllvm.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5238778&bid=17171267)

**Review by boehmseb — APPROVED** [review]

File: docs/source/vara-ts/benchbuild.rst (hunk 7 above)
Comment (vulder) [reply docs/source/vara-ts/benchbuild.rst]: good idea, done

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/experiments/wllvm.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5238778&bid=17171267)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/experiments/wllvm.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/se-passau/VaRA-Tool-Suite/pullRequest?prid=5238778&bid=17171267)

**Pull request merged**